flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/ContextResolvedFunction.java

```diff
@@ -0,0 +1,168 @@
+ * #fromCallExpression(CallExpression)} and {@link #toCallExpression(List, DataType)} allow a
+ * symmetric conversion.
+ */
+@Internal
+public class ContextResolvedFunction {
+
+    private final boolean isTemporary;
+
+    private final @Nullable FunctionIdentifier functionIdentifier;
+
+    private final FunctionDefinition functionDefinition;
+
+    public static ContextResolvedFunction permanent(
+            FunctionIdentifier functionIdentifier, FunctionDefinition functionDefinition) {
+        Preconditions.checkNotNull(
+                functionIdentifier,
+                "Function identifier should not be null for a permanent function.");
+        return new ContextResolvedFunction(false, functionIdentifier, functionDefinition);
+    }
+
+    public static ContextResolvedFunction temporary(
+            FunctionIdentifier functionIdentifier, FunctionDefinition functionDefinition) {
+        Preconditions.checkNotNull(
+                functionIdentifier,
+                "Function identifier should not be null for a temporary function.");
+        return new ContextResolvedFunction(true, functionIdentifier, functionDefinition);
+    }
+
+    public static ContextResolvedFunction anonymous(FunctionDefinition functionDefinition) {
+        return new ContextResolvedFunction(true, null, functionDefinition);
+    }
+
+    public static ContextResolvedFunction fromCallExpression(CallExpression callExpression) {
+        return new ContextResolvedFunction(
+                callExpression.isTemporary(),
+                callExpression.getFunctionIdentifier().orElse(null),
+                callExpression.getFunctionDefinition());
+    }
+
+    private ContextResolvedFunction(
+            boolean isTemporary,
+            @Nullable FunctionIdentifier functionIdentifier,
+            FunctionDefinition functionDefinition) {
+        this.isTemporary = isTemporary;
+        this.functionIdentifier = functionIdentifier;
+        this.functionDefinition =
+                Preconditions.checkNotNull(
+                        functionDefinition, "Function definition must not be null.");
+    }
+
+    public boolean isAnonymous() {
+        return functionIdentifier == null;
+    }
+
+    /** @return true if the function is temporary. An anonymous function is always temporary. */
+    public boolean isTemporary() {
+        return isTemporary;
+    }
+
+    public boolean isPermanent() {
+        return !isTemporary;
+    }
+
+    public Optional<FunctionIdentifier> getIdentifier() {
+        return Optional.ofNullable(functionIdentifier);
+    }
+
+    public FunctionDefinition getDefinition() {
+        return functionDefinition;
+    }
+
+    public String asSummaryString() {
+        if (functionIdentifier == null) {
+            return functionDefinition.toString();
+        } else {
+            return functionIdentifier.asSummaryString();
+        }
+    }
+
+    public CallExpression toCallExpression(
```

Comment:
I would keep it close to `fromCallExpression`

Reply:
not addressed, but also not insisting, more of a personal preference.

Reply:
`fromCallExpression` close to the constructor